io/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -535,11 +522,19 @@ object IO {
     override def tag = Tags.AsyncIOEffect
   }
 
-  final class Attempt[E1, E2, A] private[IO] (val value: IO[E1, A]) extends IO[E2, E1 \/ A] {
+  final class Attempt[E1, E2, A, B] private[IO] (val value: IO[E1, A],
+                                                 val err: E1 => IO[E2, B],
+                                                 val succ: A => IO[E2, B])
+      extends IO[E2, B]
+      with Function[Any, IO[Any, Any]] {
```

Comment:
Don't do the cast here. Rather, make this extend `(A => IO[E2, B])`.

We can do the cast when we push it onto the stack.